### ISSUE TITLE
Assert compiling hull shader with patch constant function

### ISSUE THREAD

**Issue by simontaylor81** [Microsoft/DirectXShaderCompiler#167]
I'm getting an assert when trying to compile a hull shader with a patch constant function specified: "must have function annotation for patch constant function". Here's a minimal repro case (compile with `/T hs_6_0`:

```
[domain("tri")]
[partitioning("fractional_odd")]
[outputtopology("triangle_cw")]
[outputcontrolpoints(3)]
[patchconstantfunc("hsConstantFunc")]
float4 main() : WORLDPOS
{
    return 0;
}

struct TessFactors
{ 
    float a[3] : SV_TessFactor;
    float i : SV_InsideTessFactor;
};
TessFactors hsConstantFunc()
{
    TessFactors output;
    output.a[0] = 0;
    output.a[1] = 0;
    output.a[2] = 0;
    output.i = 0;
    return output;
}
```

This same shader compiles successfully with fxc.

**Comment by simontaylor81** [comment]
Looks like it doesn't like that the constant patch function is declared after the main function. For whatever reason fxc is cool with this. I can probably update the shader to reverse the order so this isn't high priority. However, a good diagnostic describing the problem would be preferable over an assert :).

FWIW, the assert could be avoided by adding an early out `return` in `DxilGenerationPass::CreateDxilSignatures` after the line:

```EntryFunc->getContext().emitError("Patch constant function is not specified.");```

However, that error message is still a bit confusing.

**Comment by marcelolr** [comment]
Thanks for the report - this is something we should fix to work the way fxc does.

**Comment by antiagainst** [comment]
I think this is fixed by #748.
